[DOC-5454] Docs for Cloud Org SSO (#15397)
cockroachdb/docs: cockroachcloud/configure-cloud-org-sso.md

Members are identified by their email address. To allow members to migrate from password authentication to SSO, ensure that their email addresses in your {{ site.data.products.db }} organization match those in your identity providers. To allow your members to select from multiple SSO authentication methods, ensure that the email addresses match across all of them.

### Communicate to your members

Before you enable Cloud Organization SSO, notify your users about what to expect, such as:

- The custom login URL and when they should begin using it.
- Which authentication methods they can use and whether they have autoprovisioning enabled.
- Some members may need to be re-added to your organization:
  - All members of your {{ site.data.products.db }} organization who were using [Basic SSO](cloud-org-sso.html#basic-sso) rather than an email and password must sign in again to regain access to your organization. After signing in, members retain the same access they had before the migration.

>   - Members who are also members of other organizations must be re-added to your organization. If they sign in using an authentication method with [autoprovisioning](#autoprovisioning) enabled, they are automatically added upon successful sign-in. Otherwise, they must be re-invited. If they previously had the organizational admin role, it must be granted to them again.

Reviewer: Is "organizational admin role" the same thing as what we usually call `admin` role, or something else? If the former, then I think it needs the backticks here & elsewhere.

>   During enablement of the feature, a list of affected members is shown, and those members are also notified individually.

Reviewer: ```suggestion
During enablement of the feature, a list of affected members is shown, and those members are also notified individually.
```

      {{site.data.alerts.end}}

      Click **Next**

1. The list of default authentication methods displays. By default, **Password**, **GitHub**, **Google**, and **Microsoft** are enabled.

     To configure an authentication method, click its name and fill in the details. You can also do this later. Refer to [Configure advanced settings](#configure-advanced-settings).
     To disable an authentication method, click **Disable**. You can also do this later.

      {{site.data.alerts.callout_danger}}
      If you disable password authentication, users can no longer authenticate to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) using the passwords that were  previously stored for them, and those passwords are not retained. Cockroach Labs recommends that you thoroughly test each SSO authentication method that you enable before disabling password authentication.
      {{site.data.alerts.end}}

     Click **Next**.

1. Select **Confirm**, then click **Enable**.

> {{ site.data.products.db }} Organization SSO is enabled, and members can sign in by using the custom URL and selecting any enabled authentication method.

Reviewer: Is the feature called Cloud Organization SSO or {{ site.data.products.db }} Organization SSO? If the latter, then it needs to be consistent everywhere else.

## Update the custom URL

> After you enable Cloud Organization SSO, your users sign in use a custom URL that exposes only the authentication methods that you enable, rather than the public sign-in URL at `https://cockroachlabs.cloud`.

Reviewer: ```suggestion
After you enable Cloud Organization SSO, your users sign in at a custom URL that exposes only the authentication methods that you enable, rather than the public sign-in URL at `https://cockroachlabs.cloud`.
```

1. Go to **Organization** > **Authentication**.
1. To configure an authentication method, click its name.
1. At the top of the page, click **Edit**.
1. Click **Advanced Settings**.
1. Configure the advanced settings as desired.
1. Click **Save**.

### Allowed email domains

By default, users can access your {{ site.data.products.db }} organization from any email domain. To restrict access to a specific list of email domains:

1. Log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an organizational admin.
1. Go to **Organization** > **Authentication**.
1. To configure an authentication method, click its name.
1. At the top of the page, click **Edit**.
1. Set **Allowed Email Domains** to a comma-separated list of email domains. Each domain should begin with a `@`. To ensure that you are not locked out of {{ site.data.products.db }} while you are configuring Cloud Organization SSO, be sure to allow access to the email domain that you use to sign in.
1. Click **Save**.

### Autoprovisioning

By default, autoprovisioning is disabled. In order to log in to {{ site.data.products.db }} using SSO:

- A user's email address must exist in the SSO IdP.

> - An organizational admin user must have already invited the user to the {{ site.data.products.db }} organization.

Reviewer: ```suggestion
- An organizational admin must have already invited the user to the {{ site.data.products.db }} organization.
```

1. Log in to your IdP and gather the following information, which you will use to configure {{ site.data.products.db }} SSO:<ul><li>Issuer URL</li><li>Client ID</li><li>Client secret</li><li>Callback URL</li></ul>
1. In a separate browser, log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an organizational admin.
1. Go to **Organization** > **Authentication**.
1. Next to **Authentication Methods**, click **Add**.
1. Set **Configuration** to **OIDC (OpenID Connect)**.
1. Set **Provider Name** to a display name for the connection.
1. Click **Next**.
1. The **Provider Details** page displays.
1. To edit the connection, click **Edit**.
1. Paste the values from your IdP for the **Issuer URL**, **Client ID**, **Client Secret**, and **Callback URL**.
1. Click **Save**.
1. The authentication method has been added but is disabled. To enable it, toggle **Enable**.
1. Click **Test**. If errors are shown, edit the configuration to fix the problems and try again.
1. Optionally, [configure advanced settings](#configure-advanced-settings) for the new authentication method.

### SAML

To configure a custom SAML authentication method:

> 1. Log in to your IdP and gather the following information, which you will use to configure {{ site.data.products.db }} SSO:<ul><li>Sign-in URL</li><li>Signing certificate</li></ul>

Reviewer 2: Just went through this process to set up SSO for @MaterializeInc! It was mostly smooth, with one hiccup: I had to configure our IdP (Okta) to send along two SAML assertions.

  * `name` (unspecified name format), set to `${user.firstName} ${user.lastName}`
  * `email` (unspecified name format), set to `user.email`

It'd be great to get these requirements documented more formally, because I'm not totally sure I did it right. I figured the above out via trial and error—I tried a few different assertions and retried the "test connection" until I saw what I wanted.

Author: Thanks for this, see line 189. :)

1. In a separate browser, log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an organizational admin.
1. Go to **Organization** > **Authentication**.
1. Next to **Authentication Methods**, click **Add**.
1. Set **Configuration** to **SAML**.
1. Set **Provider Name** to a display name for the connection.
1. Click **Next**.
1. The **Provider Details** page displays.
1. To edit the connection, click **Edit**.
1. Set **Sign-in URL** to the sign-in URL from your IdP.
1. Next to **Signing Certificate**, paste the entire certificate from your IdP, including the lines `-begin certificate-` and `-end certificate`.
1. Click **Save**.
1. Click **Test**. If errors are shown, edit the configuration to fix the problems and try again.
1. Optionally, [configure advanced settings](#configure-advanced-settings) for the new authentication method.
1. Download metadata required by your IdP. Click **Download**. Open the file and make a note of the **Entity ID** and **Login URL**.
1. In the browser where you are logged in to your IdP, update the authentication configuration to use the Entity ID and Login URL from the metadata file.

> ## Require SSO

Reviewer: ```suggestion
## Require SSO

```

---

> title: Configure Cloud Organization Single Sign-On (SSO) (Preview)

Reviewer: Since the preview include is at the top, does "(Preview)" really need to be in the title?

> summary: Learn how to configure single sign-on (SSO) (Preview) for your CockroachDB Cloud organization.

Reviewer: Ditto

toc: true
docs_area: manage
---

{% include_cached feature-phases/preview-opt-in.md %}

{% include_cached cockroachcloud/sso-intro.md %}

> This page shows how to enable [Cloud Organization SSO](cloud-org-sso.html) and manage your configuration.

Reviewer: ```suggestion
This page describes how to enable [Cloud Organization SSO](cloud-org-sso.html) and manage your SSO configuration.
```

## Plan to enable Cloud Organization SSO

To ensure a smooth migration to Cloud Organization SSO, review the following information before you enable the feature.

{{site.data.alerts.callout_info}}

> After it is enabled, Cloud Organization SSO cannot be disabled. To configure Cloud Organization SSO to behave like [Basic SSO](cloud-org-sso.html#basic-sso), refer to [Emulate Basic SSO](#emulate-basic-sso)

Reviewer: ```suggestion
After it is enabled, Cloud Organization SSO **cannot** be disabled. To configure Cloud Organization SSO to behave like [Basic SSO](cloud-org-sso.html#basic-sso), refer to [Emulate Basic SSO](#emulate-basic-sso).
```

{{site.data.alerts.end}}

### Decide your custom URL

> When you [enable Cloud Organization SSO](#enable-cloud-organization-sso), you will configure a custom URL, that your users use to access {{ site.data.products.db }}. After the feature is enabled, your users must use that page to sign in. Your custom URL must be unique within {{ site.data.products.db }}. Communicate the custom URL to your users ahead of time.

Reviewer: Recommending some bold here, but up to you.

```suggestion
When you [enable Cloud Organization SSO](#enable-cloud-organization-sso), you will configure a custom URL through which your users can access {{ site.data.products.db }}. After the feature is enabled, your users must use that page to sign in. Your custom URL must be unique within {{ site.data.products.db }}. **Communicate this custom URL to your users ahead of time.**
```

> To change your custom URL after Cloud Organization SSO has been enabled, contact your account team.

Reviewer: ```suggestion
To change your custom URL after enabling Cloud Organization SSO, contact your account team.
```

### Decide which authentication methods to enable

All enabled authentication methods appear on your custom URL and are available to your users. By default, when you enable Cloud Organization SSO, the following authentication methods are enabled by default:

> - Email and password

Reviewer: Is this what it's called in the UI?

> - Email and password

Author: ```suggestion
- Password
```

> - Github

Reviewer: ```suggestion
- GitHub
```

- Google
- Microsoft

In addition, you can create authentication methods that connect to your identity provider using the [Security Access Markup Language (SAML)](https://en.wikipedia.org/wiki/Security_Assertion_Markup_Language) and [OpenID Connect (OIDC)](https://openid.net/connect/) identity protocols.

> Members are identified by their email address. To allow your members to select from multiple SSO authentication methods, ensure that the email addresses match across all of them. To allow members to migrate from password authentication to SSO, ensure that their email addresses match those in your identity providers.

Reviewer: Not quite sure what "ensure that the email addresses match across all of them" means.

Reviewer 3: If a user has a email/password account as abhinav@cockroachlabs.com, then when the admin enabled SSO using Okta, it must be ensured that it's the same email in there as well (and vice-versa) to allow for seamless switch between enabled methods. Otherwise if the email/password account is setup as abhinav@gmail.com, that would be treated as a different user altogether within the org.

> Members are identified by their email address. To allow your members to select from multiple SSO authentication methods, ensure that the email addresses match across all of them. To allow members to migrate from password authentication to SSO, ensure that their email addresses match those in your identity providers.

Author: ```suggestion
Members are identified by their email address. To allow members to migrate from password authentication to SSO, ensure that their email addresses in your {{ site.data.products.db }} organization match those in your identity providers. To allow your members to select from multiple SSO authentication methods, ensure that the email addresses match across all of them. 
```

### Prepare for migration of existing members

> Keep the following in mind and communicate it to your members:

Reviewer: See my following comment.

Otherwise, same comments on this section as in cloud-org-sso.md. And perhaps this should be in an include across both docs.

- If your organization includes members who are also members of other organizations, and this is the first of their organizations to enable Cloud Organization SSO, those members must be re-added to your organization. If they sign in using an authentication method with [autoprovisioning](#autoprovisioning) enabled, they are automatically added upon successful sign-in. **However, those who were previously admins must be granted the role again**.
- After you enable Cloud Organization SSO, all other members of your {{ site.data.products.db }} organization who were using [Basic SSO](cloud-org-sso.html#basic-sso) rather than an email and password must sign in again to regain access to your organization. After signing in, members retain the same access they had before the migration.

During enablement of the feature, a list of affected members is shown, and those members are also notified individually.

### Communicate to your members

> Communicate to your users about:

Reviewer: It seems easiest to fold the 2 bullets in the preceding header into this section. This is assuming everything in the preceding header describes a point that needs to be communicated to the user. In any case, the two headings somewhat overlap as written.

Author: Good idea. Done.

- For members who are admins in your organization and are also members of other organizations, they may need to be re-added to your organization and re-assigned the admin role. Refer to [Prepare for migration of existing members](#prepare-for-migration-of-existing-members).

## Enable Cloud Organization SSO

To enable Cloud Organization SSO:

1. Log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.
1. Go to **Organization** > **Authentication**.
1. Next to **Enable Authentication**, click **Enable**.
1. In the dialog, configure the custom URL your users will use to sign in. This value must be unique across {{ site.data.products.db }}. For more details, refer to [Update the custom URL](#update-the-custom-url).

      {{site.data.alerts.callout_info}}
      If you previously updated the custom URL from **Organization** > **Authentication** but did not enable Cloud Organization SSO, you are not prompted to configure it again. The value you have already configured is automatically used during enablement. To edit the custom URL again, click **Back** in the dialog.
      {{site.data.alerts.end}}

      Click **Next**

> 1. The list of default authentication methods displays. By default, **Email**, **Github**, **Google**, and **Microsoft** are enabled.

Reviewer: ```suggestion
1. The list of default authentication methods displays. By default, **Email**, **GitHub**, **Google**, and **Microsoft** are enabled.
```

     To configure an authentication method, click its name and fill in the details. You can also do this later. Refer to [Configure advanced settings](#configure-advanced-settings).
     To disable an authentication method, click **Disable**. You can also do this later.

      {{site.data.alerts.callout_danger}}
      If you disable password authentication, users can no longer authenticate to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) using the passwords that were  previously stored for them, and those passwords are not retained. Cockroach Labs recommends that you thoroughly test each SSO authentication method that you enable before disabling password authentication.
      {{site.data.alerts.end}}

     Click **Next**.

1. Select **Confirm**, then click **Enable**.

{{ site.data.products.db }} Organization SSO is enabled, and members can sign in by using the custom URL and selecting any enabled authentication method.

## Update the custom URL

> After you enable Cloud Organization SSO, your users sign in use a custom URL that exposes only the authentication methods that you enable, rather than the [public sign-in URL](https://cockroachlabs.cloud).

Reviewer: ```suggestion
After you enable Cloud Organization SSO, your users sign in use a custom URL that exposes only the authentication methods that you enable, rather than the public sign-in URL at `https://cockroachlabs.cloud`.
```

Author: Good point.

To update the custom URL after Cloud Organization SSO is enabled, contact your account team.

## Enable or disable an authentication method

> When you enable an authentication method, users can begin authenticating to {{ site.data.products.db }} using that authentication method. A user can log in using any enabled authentication method. When they successfully sign in with a new authentication method for the first time, this becomes their new default method. You can optionally [configure advanced settings](#configure-advanced-settings) for each enabled authentication method.

Reviewer 3: ```suggestion
When you enable an authentication method, users can begin authenticating to {{ site.data.products.db }} using that authentication method. A user can log in using any enabled authentication method, as long as the email address for a user is the same across all enabled methods. When they successfully sign in with a new authentication method for the first time, this becomes their new default method. You can optionally [configure advanced settings](#configure-advanced-settings) for each enabled authentication method.
```

> When you disable an authentication method, users who are no longer associated with any enabled authentication method can no longer sign in and must be provisioned in an identity provider that is associated with an enabled method.

Reviewer 3: ```suggestion
When you disable an authentication method, users who are no longer associated with the other enabled authentication method(s) can no longer sign in and must be provisioned in an identity provider that is associated with one of those other enabled methods.
```

1. To enable or disable the authentication method, toggle **Enable**.
1. If desired, configure [advanced settings](#configure-advanced-settings) for the authentication method.
1. Click **Save**.

## Configure advanced settings

The following sections describe the advanced settings you can configure for an SSO authentication method. To configure an authentication method:

1. Log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.
1. Go to **Organization** > **Authentication**.
1. To configure an authentication method, click its name.
1. At the top of the page, click **Edit**.
1. Click **Advanced Settings**.
1. Configure the advanced settings as desired.
1. Click **Save**.

### Allowed email domains

By default, users can access your {{ site.data.products.db }} organization from any email domain. To restrict access to a specific list of email domains:

1. Log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.
1. Go to **Organization** > **Authentication**.
1. To configure an authentication method, click its name.
1. At the top of the page, click **Edit**.

> 1. Set **Allowed Email Domains** to a comma-separated list of email domains. domains in CIDR format. Each domain should begin with a `@`. To ensure that you are not locked out of {{ site.data.products.db }} while you are configuring Cloud SSO, be sure to allow access to the email domain that you use to sign in.

Reviewer: ```suggestion
1. Set **Allowed Email Domains** to a comma-separated list of email domains. domains in CIDR format. Each domain should begin with a `@`. To ensure that you are not locked out of {{ site.data.products.db }} while you are configuring Cloud Organization SSO, be sure to allow access to the email domain that you use to sign in.
```

Reviewer 3: ```suggestion
1. Set **Allowed Email Domains** to a comma-separated list of email domains. Each domain should begin with a `@`. To ensure that you are not locked out of {{ site.data.products.db }} while you are configuring Cloud Organization SSO, be sure to allow access to the email domain that you use to sign in.
```

Author: Combined into Ryan's suggestion.

1. Click **Save**.

### Autoprovisioning

By default, autoprovisioning is disabled. In order to log in to {{ site.data.products.db }} using SSO:

> - A user's email address must exist in the SSO IdP, **and**

Reviewer: Note the added line break to correctly format the list

```suggestion

- A user's email address must exist in the SSO IdP.
```

- An admin user must have already invited the user to the {{ site.data.products.db }} organization.

Autoprovisioning allows members to access your organization without an invitation. Users are assigned the [Developer role](/docs/cockroachcloud/console-access-management.html#developer) by default. If you enable autoprovisioning, Cockroach Labs recommends that you also limit the [allowed email domains](#allowed-email-domains) for the authentication method.

> To enable auto-provisioning for an SSO authentication method:

Author: ```suggestion
To enable autoprovisioning for an SSO authentication method:
```

1. Log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.
1. Go to **Organization** > **Authentication**.
1. Click the name of an authentication method.
1. Click **Advanced Settings**.
1. Next to **Autoprovisioning**, click **Enable**.

## Add a custom authentication method

You can add a custom authentication method to connect to any IdP that supports [Security Access Markup Language (SAML)](https://en.wikipedia.org/wiki/Security_Assertion_Markup_Language) or [OpenID Connect (OIDC)](https://openid.net/connect/).

### OIDC

To configure a custom OIDC authentication method:

> 1. Log in to your IdP and gather the following information, which you will use to configure {{ site.data.products.db }} SSO:<ul><li>Issuer URL</li><li>Client ID</li><li>Client secret</li><li>Callback URL</li></ul>

Reviewer: Yikes, perhaps we should log a docs tooling request to have this work in Markdown (if possible).

Author: There are some contexts that I just can't get sub-bullets to work. cc/ @nickvigilante

> 1. In a separate browser, log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.

Reviewer: Should this be `admin`?

Reviewer 3: Org Admin

> 1. In a separate browser, log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.

Author: ```suggestion
1. In a separate browser, log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an organizational admin.
```

1. Go to **Organization** > **Authentication**.
1. Next to **Authentication Methods**, click **Add**.
1. Set **Configuration** to **OIDC (OpenID Connect)**.
1. Set **Provider Name** to a display name for the connection.
1. Click **Next**.
1. The **Provider Details** page displays.
1. To edit the connection, click **Edit**.
1. Paste the values from your IdP for the **Issuer URL**, **Client ID**, **Client Secret**, and **Callback URL**.
1. Click **Save**.
1. The authentication method has been added but is disabled. To enable it, toggle **Enable**.
1. Click **Test**. If errors are shown, edit the configuration to fix the problems and try again.
1. Optionally, [configure advanced settings](#configure-advanced-settings) for the new authentication method.

### SAML

To configure a custom SAML authentication method:

1. Log in to your IdP and gather the following information, which you will use to configure {{ site.data.products.db }} SSO:<ul><li>Sign-in URL</li><li>Signing certificate</li></ul>

> 1. In a separate browser, log in to [{{ site.data.products.db }} Console](https://cockroachlabs.cloud) as an admin user.

Reviewer: Ditto re: admin

Author: Fixed everywhere.

1. Set **Sign-in URL** to the sign-in URL from your IdP.
1. Next to **Signing Certificate**, paste the entire certificate from your IdP, including the lines `-begin certificate-` and `-end certificate`.
1. Click **Save**.
1. Click **Test**. If errors are shown, edit the configuration to fix the problems and try again.
1. Optionally, [configure advanced settings](#configure-advanced-settings) for the new authentication method.
1. Download metadata required by your IdP. Click **Download**. Open the file and make a note of the **Entity ID** and **Login URL**.
1. In the browser where you are logged in to your IdP, update the authentication configuration to use the Entity ID and Login URL from the metadata file.

## Require SSO
To begin enforcing a requirement to sign in using SSO:

1. [Enable Cloud Organization SSO](#enable-cloud-organization-sso).
1. [Disable SSO authentication methods](#enable-or-disable-an-authentication-method) that you don't need and optionally [configure advanced settings](#configure-advanced-settings) for those that are enabled.
1. After thorough testing, [disable email authentication](#enable-or-disable-an-authentication-method).

## Emulate Basic SSO

> After Cloud Organization SSO is enabled, it can't be disabled. To emulate the behavior of Basic SSO:

Reviewer: ```suggestion
After Cloud Organization SSO is enabled, it cannot be disabled. To emulate the behavior of Basic SSO:
```

1. Enable the following authentication methods:
      - Password

>       - Github

Reviewer: ```suggestion
      - GitHub
```

      - Google
      - Microsoft

> 1. For Github, Google, and Microsoft authentication methods, allow all email domains and disable autoprovisioning.

Reviewer: ```suggestion
1. For GitHub, Google, and Microsoft authentication methods, allow all email domains and disable autoprovisioning.
```
